### PR TITLE
Add support for Happy Eyeballs DNS Address Resolution

### PR DESCRIPTION
**Pull request by nicoburniske**
/claim #1996 

Currently facing unit test failure. NettyConnectionPoolSpec tests are not terminating

### CODE REVIEW

File: zio-http/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala [hunk 1]
```diff
@@ -114,8 +108,107 @@ object NettyConnectionPool {
...
+  private def executeHappyEyeballs[A](tasks: List[Task[A]], delay: Duration): Task[A] =
+    tasks match {
+      // What should this error be?
+      case Nil                => ZIO.fail(new IllegalStateException("No tasks left."))
+      case task :: Nil        => task
+      case task :: otherTasks =>
+        Promise.make[Nothing, Unit].flatMap { failurePromise =>
+          val taskWithFailure = task.onError(_ => failurePromise.complete(ZIO.unit).ignore)
+          val continue        = failurePromise.await.timeout(delay).ignore *> executeHappyEyeballs(otherTasks, delay)
+          taskWithFailure.race(continue)
```
Comment (nicoburniske) [inline zio-http/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala]: In the NettyConnectionPoolSpec tests `taskWithFailure` is completing, but the function does not terminate. I am not sure why the race on the right `continue` is not being interrupted. I am sure I'm doing something wrong but I don't understand what exactly

File: zio-http/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala [hunk 2]
```diff
@@ -114,8 +108,107 @@ object NettyConnectionPool {
       _             <- NettyFutureExecutor.executed(channelFuture)
       result        <- ZIO.attempt(channelFuture.channel())
     } yield result
+
+    for {
+      resolvedHosts <- dnsResolver.resolve(location.host)
+      sortedHosts    = sortHostsHappyEyeballs(resolvedHosts)
+      createChannels = sortedHosts.map(createChannel).toList
```
Comment (nicoburniske) [inline zio-http/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala]: If I replace this line with 

`sortedHosts.map(createChannel).map(_.debug("create channel")).toList`

the create channel log comes out as a success, but the function still fails to terminate.

**Comment by jdegoes** [comment]
I know we need to take another look at this in a session, but also, it's important that we make sure the current code path is not negatively affected, performance-wise.

**Comment by jdegoes** [comment]
I will close for now, please re-open when you are ready for a final review!

**Pull request closed without merging**